src/script/api/script_goal.cpp

```diff
@@ -28,29 +28,32 @@
 	return ::Goal::IsValidID(goal_id);
 }
 
-/* static */ ScriptGoal::GoalID ScriptGoal::New(ScriptCompany::CompanyID company, Text *goal, GoalType type, SQInteger destination)
+/* static */ bool ScriptGoal::IsValidGoalDestination(ScriptCompany::CompanyID company, GoalType type, SQInteger destination)
 {
-	CCountedPtr<Text> counter(goal);
-
-	EnforceDeityMode(GOAL_INVALID);
-	EnforcePrecondition(GOAL_INVALID, goal != nullptr);
-	const std::string &text = goal->GetEncodedText();
-	EnforcePreconditionEncodedText(GOAL_INVALID, text);
-	EnforcePrecondition(GOAL_INVALID, company == ScriptCompany::COMPANY_INVALID || ScriptCompany::ResolveCompanyID(company) != ScriptCompany::COMPANY_INVALID);
-
 	CompanyID c = (::CompanyID)company;
 	if (company == ScriptCompany::COMPANY_INVALID) c = INVALID_COMPANY;
 	StoryPage *story_page = nullptr;
 	if (type == GT_STORY_PAGE && ScriptStoryPage::IsValidStoryPage((ScriptStoryPage::StoryPageID)destination)) story_page = ::StoryPage::Get((ScriptStoryPage::StoryPageID)destination);
-
-	EnforcePrecondition(GOAL_INVALID, (type == GT_NONE && destination == 0) ||
+	return (type == GT_NONE && destination == 0) ||
 			(type == GT_TILE && ScriptMap::IsValidTile(destination)) ||
 			(type == GT_INDUSTRY && ScriptIndustry::IsValidIndustry(destination)) ||
 			(type == GT_TOWN && ScriptTown::IsValidTown(destination)) ||
 			(type == GT_COMPANY && ScriptCompany::ResolveCompanyID((ScriptCompany::CompanyID)destination) != ScriptCompany::COMPANY_INVALID) ||
-			(type == GT_STORY_PAGE && story_page != nullptr && (c == INVALID_COMPANY ? story_page->company == INVALID_COMPANY : story_page->company == INVALID_COMPANY || story_page->company == c)));
+			(type == GT_STORY_PAGE && story_page != nullptr && (c == INVALID_COMPANY ? story_page->company == INVALID_COMPANY : story_page->company == INVALID_COMPANY || story_page->company == c));
```

Comment:
Could this use the new `IsValidGoalDestination` function?

Reply:
It's something I considered. The original author had this code written twice like this. Without understanding the original intention I followed suit.

If I had to guess, the intention seems to be to use functions like `ScriptMap::IsValidTile` instead of `::IsValidTile`, but these functions in `ScriptMap` mostly just call the global functions anyway. Maybe the intention is to try and make sure the checks obey the logic exposed to the script api. 

If you agree this isn't a concern I can replace this with  a call to `Goal::IsValidGoalDestination`